File: leetcode/282.py
class Solution(object):
    def addOperators(self, num, target):
        def solve(target, pos, negate, prod):
            ans = []
            n = len(num)
            for i in xrange(pos, n):
                if i > pos and num[pos] == "0": break
                if i == n - 1:
                    if negate * prod * int(num[pos : i + 1]) == target: ans.extend([num[pos : i + 1]])
                    break
                add_expr = solve(target - prod * negate * long(num[pos : i + 1]), i + 1, 1, 1)
                ans.extend([num[pos : i + 1] + "+" + e for e in add_expr])
                sub_expr = solve(target - prod * negate * long(num[pos : i + 1]), i + 1, -1, 1)     
                ans.extend([num[pos : i + 1] + "-" + e for e in sub_expr])
                mul_expr = solve(target, i + 1, 1, prod * negate * long(num[pos : i + 1]))
                ans.extend([num[pos : i + 1] + "*" + e for e in mul_expr])
            return ans
        return solve(target, 0, 1, 1)

File: leetcode/282.cpp
class Solution {
public:
    void dfs(vector<string>& ans, string now, string num, long long last, long long cur, int target) {
        if (!num.size() && cur == target) {
            ans.push_back(now);
            return;
        }
        for (int i = 1; i <= num.size(); ++i) {
            string t = num.substr(0, i);
            if (t.size() > 1 && t[0] == '0') continue;
            string nxt = num.substr(i);
            if (now.size() > 0) {
                dfs(ans, now + "+" + t, nxt, stoll(t), cur + stoll(t), target);
                dfs(ans, now + "-" + t, nxt, -stoll(t), cur - stoll(t), target);
                dfs(ans, now + "*" + t, nxt, last * stoll(t), (cur - last) + last * stoll(t), target);
            }
            else dfs(ans, t, nxt, stoll(t), stoll(t), target);
        }
    }
    vector<string> addOperators(string num, int target) {
        vector<string> ans;
        dfs(ans, "", num, 0, 0, target);
        return ans;
    }
};